apollo-client/src/main/java/com/ctrip/framework/apollo/internals/K8sConfigMapConfigRepository.java

```diff
@@ -0,0 +1,301 @@
+
+import com.ctrip.framework.apollo.kubernetes.KubernetesManager;
+import com.ctrip.framework.apollo.build.ApolloInjector;
+import com.ctrip.framework.apollo.core.ConfigConsts;
+import com.ctrip.framework.apollo.core.utils.DeferredLoggerFactory;
+import com.ctrip.framework.apollo.core.utils.StringUtils;
+import com.ctrip.framework.apollo.enums.ConfigSourceType;
+import com.ctrip.framework.apollo.exceptions.ApolloConfigException;
+import com.ctrip.framework.apollo.tracer.Tracer;
+import com.ctrip.framework.apollo.tracer.spi.Transaction;
+import com.ctrip.framework.apollo.util.ConfigUtil;
+import com.ctrip.framework.apollo.util.ExceptionUtil;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.slf4j.Logger;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * @author dyx1234
+ */
+public class K8sConfigMapConfigRepository extends AbstractConfigRepository
+        implements RepositoryChangeListener {
+    private static final Logger logger = DeferredLoggerFactory.getLogger(K8sConfigMapConfigRepository.class);
+    private final String namespace;
+    private String configMapName;
+    private String configMapKey;
+    private final String k8sNamespace;
+    private final ConfigUtil configUtil;
+    private final KubernetesManager kubernetesManager;
+    private volatile Properties configMapProperties;
+    private volatile ConfigRepository upstream;
+    private volatile ConfigSourceType sourceType = ConfigSourceType.CONFIGMAP;
+
+    /**
+     * Constructor
+     *
+     * @param namespace the namespace
+     */
+    public K8sConfigMapConfigRepository(String namespace) {
+        this(namespace, (ConfigRepository) null);
+    }
+
+    public K8sConfigMapConfigRepository(String namespace, KubernetesManager kubernetesManager) {
+        this.namespace = namespace;
+        configUtil = ApolloInjector.getInstance(ConfigUtil.class);
+        k8sNamespace = configUtil.getK8sNamespace();
+        this.kubernetesManager = kubernetesManager;
+
+        this.setConfigMapKey(configUtil.getCluster(), namespace);
+        this.setConfigMapName(configUtil.getAppId(), false);
+        this.setUpstreamRepository(upstream);
+    }
+
+    public K8sConfigMapConfigRepository(String namespace, ConfigRepository upstream) {
+        this.namespace = namespace;
+        configUtil = ApolloInjector.getInstance(ConfigUtil.class);
+        kubernetesManager = ApolloInjector.getInstance(KubernetesManager.class);
+        k8sNamespace = configUtil.getK8sNamespace();
+
+        this.setConfigMapKey(configUtil.getCluster(), namespace);
+        this.setConfigMapName(configUtil.getAppId(), false);
+        this.setUpstreamRepository(upstream);
+    }
+
+    public String getConfigMapKey() {
+        return configMapKey;
+    }
+
+    public String getConfigMapName() {
```

Comment:
We could make it package private as it's only used in the unit test.
```suggestion
    String getConfigMapName() {
```